File: upgrade_db_version.rb
require 'mysql2'
 
HOST = '127.0.0.1'
USERNAME = 'root'
PASSWORD = '123456'
DB_CHANGE_PATH = './db_changes/'

client = Mysql2::Client.new(
    :host     => HOST, 
    :username => USERNAME,      
    :password => PASSWORD,
    :flags =>  Mysql2::Client::MULTI_STATEMENTS,
    )


class C
    def initialize(c)
       @client = c
    end

    def query(sql)
        @client.query(sql)
        while @client.next_result
            result = @client.store_result
        end
    end
    
    def update (db,path)
        @client.query("USE information_schema")
    
        sql = "SELECT `TABLE_NAME` FROM `TABLES` WHERE `TABLE_NAME` = 'db_version' AND `TABLE_SCHEMA` = '#{db}'"
        results = @client.query(sql)
    
        @client.query("USE #{db}")
    
        if results.size == 0 
            sql = "CREATE TABLE `db_version` ( `version` INT ) ENGINE 'InnoDb' CHARACTER SET 'utf8' COLLATE 'utf8_general_ci'"
            @client.query(sql)
        end
    
        sql = "SELECT `version` FROM `db_version`"
        results = @client.query(sql)
    
        if results.size == 0 
            sql = "INSERT INTO `db_version` VALUES (0)"
            @client.query(sql)
            latest_version = 0
        else
            results.each do |row|
                latest_version = row['version']
            end
        end
    
        changes_array = Array.new()
        changes_hash = Hash.new()
    
        Dir.foreach(path) do |entry|
            if entry != '.' and entry != '..'
                extname = File.extname(entry)
                version = entry.delete(extname)
                version = version.gsub('-','')
                version = version.to_i()
                if version > latest_version
                    changes_array.push(version) 
                    changes_hash[version] = entry
                end
            end
        end
    
        if changes_array.size == 0 
            return
        end
        
        changes_array.sort!()
        begin
            @client.query("BEGIN") 
                changes_array.each do |version|
                    filename = changes_hash[version]
                    puts "[#{db}]=>"+filename
                    script = File.read(path + "/" + filename)
                    # puts script
                    sql = ""
                    if File.extname(filename) == ".rb"
                        eval(script)
                    else
                        self.query(script)
                    end
                    latest_version = version
                end
            sql = "UPDATE `db_version` SET `version` = #{latest_version}" 
            @client.query(sql)
            @client.query("COMMIT")
        rescue => e
            @client.query("ROLLBACK")
            puts e.inspect()
            raise
        end
        
        puts "数据库[#{db}]已更新至版本:#{latest_version}"
    end

 end
 

c = C.new(client)

Dir.foreach(DB_CHANGE_PATH) do |entry|
    if entry != '.' and entry != '..'
       path = "#{DB_CHANGE_PATH}" + entry
       if File.directory?(path)
          c.update(entry,path)
       end
    end
end

puts "所有数据库已更新至最新版本"

File: upgrade_db_version.js
const mysql = require('mysql2/promise');
const readline = require('readline');
const fs = require('fs');

let HOST = "127.0.0.1";
let USER = "root";
let PASSWORD = "123456";
let DB_CHANGE_PATH = "./db_changes";
let DB = "";

async function upgrade(conn, db_name, path) {
    await conn.query("USE information_schema");

    let result = await conn.query("SELECT `TABLE_NAME` FROM `TABLES` WHERE `TABLE_NAME` = 'db_version' AND `TABLE_SCHEMA` = ?", [db_name]);

    await conn.query("USE " + db_name);

    if (result[0].length == 0) {
        await conn.query(
            "CREATE TABLE `db_version` ( `version` INT,`is_execute` INT,PRIMARY KEY (`version`))  ENGINE 'InnoDb' CHARACTER SET 'utf8' COLLATE 'utf8_general_ci'"
        );
    }

    let changes_array = new Array();
    let changes_map = new Map();
    const files = fs.readdirSync(path)
    for (let f of files) {
        let version = f.split(".")[0];
        if (version) {
            changes_array.push(version);
            changes_map[version] = f;
        }
    }
    changes_array.sort();

    let db_version = new Map();
    result = await conn.query("SELECT * FROM `db_version`");
    for (let row of result[0]) {
        db_version[row.version] = row.is_execute
    }

    for (let version of changes_array) {
        if (db_version[version]) {
            continue
        }
        let file_name = changes_map[version];
        console.log("[" + db_name + "]=>" + file_name);
        let script = fs.readFileSync(path + "/" + file_name, "utf-8");
        if (script.charCodeAt(0) === 0xFEFF) {
            script = script.slice(1);
        }
        await conn.query(script);
        await conn.query("INSERT INTO `db_version` VALUES (?,?)", [version, 1]);
    }

    result = await conn.query("SELECT MAX(`version`) FROM `db_version`");
    let latest_version = result[0][0]['MAX(`version`)']
    console.log("数据库[" + db_name + "]已更新至版本:" + latest_version);
}

function Ask(query) {
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    })
    return new Promise(resolve => rl.question(query, ans => {
        rl.close();
        resolve(ans);
    }))
}

function ParseFlag(argv) {
    for (let arg of argv) {
        let flag = arg.split('=')
        switch (flag[0]) {
            case "DB":
                DB = flag[1]
                break
            case "HOST":
                HOST = flag[1]
                break
            case "USER":
                USER = flag[1]
                break
            case "PASSWORD":
                PASSWORD = flag[1]
                break
            case "DB_CHANGE_PATH":
                DB_CHANGE_PATH = flag[1]
                break
        }
    }
}

(async function main() {
    ParseFlag(process.argv)
    if (!DB) {
        DB = process.env.MY_DB
    }
    if (!DB) {
        DB = await Ask("请输入你的数据库名:")
    }
    let dbs = DB.split(',')
    const conn = await mysql.createConnection({ host: HOST, user: USER, password: PASSWORD, multipleStatements: true });
    try {
        await conn.beginTransaction();
        for (let db of dbs) {
            await upgrade(conn, db, DB_CHANGE_PATH);
        }
        await conn.commit();
    } catch (err) {
        await conn.rollback();
        throw err;
    } finally {
        await conn.end();
    }
})();
